core/src/main/java/org/apache/calcite/util/format/PostgresqlDateTimeFormatter.java

```diff
@@ -0,0 +1,660 @@
+
+  /**
+   * The format patterns that are supported. Order is very important, since some patterns
+   * are prefixes of other patterns.
+   */
+  @SuppressWarnings("TemporalAccessorGetChronoField")
+  private static final FormatPattern[] FORMAT_PATTERNS = new FormatPattern[] {
+      new NumberFormatPattern(
+          dt -> {
+            final int hour = dt.get(ChronoField.HOUR_OF_AMPM);
+            return String.format(Locale.ROOT, "%02d", hour == 0 ? 12 : hour);
+          },
+          "HH12"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%02d", dt.getHour()),
+          "HH24"),
+      new NumberFormatPattern(
+          dt -> {
+            final int hour = dt.get(ChronoField.HOUR_OF_AMPM);
+            return String.format(Locale.ROOT, "%02d", hour == 0 ? 12 : hour);
+          },
+          "HH"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%02d", dt.getMinute()),
+          "MI"),
+      new NumberFormatPattern(
+          dt -> Integer.toString(dt.get(ChronoField.SECOND_OF_DAY)),
+          "SSSSS", "SSSS"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%02d", dt.getSecond()),
+          "SS"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%03d", dt.get(ChronoField.MILLI_OF_SECOND)),
+          "MS"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%06d", dt.get(ChronoField.MICRO_OF_SECOND)),
+          "US"),
+      new NumberFormatPattern(
+          dt -> Integer.toString(dt.get(ChronoField.MILLI_OF_SECOND) / 100),
+          "FF1"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%02d", dt.get(ChronoField.MILLI_OF_SECOND) / 10),
+          "FF2"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%03d", dt.get(ChronoField.MILLI_OF_SECOND)),
+          "FF3"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%04d", dt.get(ChronoField.MICRO_OF_SECOND) / 100),
+          "FF4"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%05d", dt.get(ChronoField.MICRO_OF_SECOND) / 10),
+          "FF5"),
+      new NumberFormatPattern(
+          dt -> String.format(Locale.ROOT, "%06d", dt.get(ChronoField.MICRO_OF_SECOND)),
+          "FF6"),
+      new StringFormatPattern(
+          dt -> dt.getHour() < 12 ? "AM" : "PM",
+          "AM", "PM"),
+      new StringFormatPattern(
+          dt -> dt.getHour() < 12 ? "am" : "pm",
+          "am", "pm"),
+      new StringFormatPattern(
+          dt -> dt.getHour() < 12 ? "A.M." : "P.M.",
+          "A.M.", "P.M."),
+      new StringFormatPattern(
+          dt -> dt.getHour() < 12 ? "a.m." : "p.m.",
+          "a.m.", "p.m."),
+      new NumberFormatPattern(dt -> {
+        final String formattedYear = String.format(Locale.ROOT, "%0,4d", dt.getYear());
+        if (formattedYear.length() == 4 && formattedYear.charAt(0) == '0') {
```

Comment:
this implies that years above 10000 wont' be correct. Maybe that's within spec.

Reply:
`%0,4d` means to print out at least 4 digits and formatted with commas. A value like `12345` will become `12,345`.

There is the special case, where we get a 4 digit string back that has at least one leading 0. In this case, Java will not add the comma. For example `1` will become `0001`. Need to insert the comma to match PostgreSQL behaviour.